crates/storage/api.rs

```diff
@@ -1,361 +1,142 @@
-            return Ok(None);
-        };
-        self.get_block_by_hash(block_hash).await
-    }
-
-    // Get the canonical block hash for a given block number.
-    async fn get_canonical_block_hash(
-        &self,
-        block_number: BlockNumber,
-    ) -> Result<Option<BlockHash>, StoreError>;
-
-    /// Stores the chain configuration values, should only be called once after reading the genesis file
-    /// Ignores previously stored values if present
-    async fn set_chain_config(&self, chain_config: &ChainConfig) -> Result<(), StoreError>;
+//! # Storage Backend API
+//!
+//! This module provides a thin, minimal interface for storage backends:
+//!
+//! - Thin: Minimal set of operations that databases must provide
+//! - Simple: Avoids type-system complexity and focuses on core functionality
+//!
+//! Rather than implementing business logic in each database backend, this API
+//! provides low-level primitives that higher-level code can build upon.
+//! This eliminates code duplication and makes adding new database backends trivial.
+//!
+//! The API differentiates between three types of database access:
+//!
+//! - Read-only transactions ([`StorageRoTx`])
+//! - Read-write transactions ([`StorageRwTx`])
+//! - Locked snapshots ([`StorageLocked`]): Persistent read-only views, righ now it's
+//!   only used in snapsync stage.
+
+use crate::error::StoreError;
+use std::{fmt::Debug, path::Path};
+
+pub use tables::LIST as TABLES;
+
+/// Type alias for the result of a prefix iterator.
+/// Uses Box<[u8]> instead of Vec<u8> to avoid unnecessary allocations
+/// when the data comes directly from the database.
+pub type PrefixResult = Result<(Box<[u8]>, Box<[u8]>), StoreError>;
+
+/// Configuration options for table creation.
+#[derive(Debug, Clone)]
+pub struct TableOptions {
+    /// Whether the table supports duplicate keys, this means that multiple values can be stored for the same key.
+    /// This is useful for certain indexing scenarios but not supported by all backends.
+    pub dupsort: bool,
+}
 
-    /// Update earliest block number
-    async fn update_earliest_block_number(
-        &self,
-        block_number: BlockNumber,
-    ) -> Result<(), StoreError>;
+/// This trait provides a minimal set of operations required from a database backend.
+/// Implementations should focus on providing efficient access to the underlying storage
+/// without implementing business logic.
+pub trait StorageBackend: Debug + Send + Sync {
+    /// Opens a storage backend at the specified path.
+    fn open(path: impl AsRef<Path>) -> Result<Self, StoreError>
+    where
+        Self: Sized;
 
-    /// Obtain earliest block number
-    async fn get_earliest_block_number(&self) -> Result<Option<BlockNumber>, StoreError>;
+    /// Creates a new table, allowing to specify [`TableOptions`].
+    fn create_table(&self, name: &'static str, options: TableOptions) -> Result<(), StoreError>;
```

Comment:
Do we want to allow arbitrary table creation? I think it's safe to assume the backend automatically configures the needed tables for `ethrex` to use. Otherwise, we'll need to find some way to allow for tuning the performance of each backend without adding too much complexity to the whole thing.